### ISSUE TITLE
Router should support request body types other than hyper::Body

### ISSUE THREAD

**Issue by sfackler** [hyperium/tonic#905]
The generated server Service implementations already support arbitrary request body types, but Router is currently hardcoded to work with hyper::Body directly. It'd be great if that were relaxed to work with any suitable body type.

It may be doable in a backwards compatible way with a new default type parameter? I'm going to give it a go.

**Comment by LucioFranco** [comment]
Agreed we can make this change. I believe we should do this in conjunction with https://github.com/hyperium/tonic/pull/830

**Comment by LucioFranco** [comment]
I am going to remove this from 0.7 because it will be quite challenging to get this change in right now and I would actually like to refactor the `transport` module in the next release (aka remove it and re do it).

**Comment by davidpdrsn** [comment]
Yes that's what I meant.

**Comment by sfackler** [comment]
So then it's totally possible to create a Hyper service that changes the request body type and then feeds that modified body type into the Tonic Router.

**Comment by tobz** [comment]
This would be extremely helpful for a particular issue I'm working on. As @sfackler points out, the major sticking point seems to be where `tonic` meets `axum`. Without this, the only current workaround is doing very goofy stuff like getting a channel-backed `hyper::Body`, even if you only needed to do something like count the number of bytes per data chunk emitted by the body, etc.

I'd be interested in trying to work on and submit a PR for this. IS this something you'd actually accept, though, given the thoughts/plans around breaking `tonic` apart into more composable chunks?
